src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs

```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+
+namespace System.Text.Json
+{
+    internal sealed class JsonSnakeCaseNamingPolicy : JsonNamingPolicy
+    {
+        public override string ConvertName(string name)
+        {
+            if (string.IsNullOrEmpty(name))
+                return name;
+
+            StringBuilder builder = new StringBuilder(name.Length + Math.Min(2, name.Length / 5));
```

Comment:
Would this benefit from `ValueStringBuilder` with a `stackalloc` buffer (alocated with a size cap)?

Reply:
Sure, but in that case `ValueStringBuilder` should be referenced by the project since the struct isn't public yet (see #28379). If it's okay, I will do it (:

/cc @stephentoub @ahsonkhan

Reply:
Since `ValueStringBuilder` is internal (within corelib), I wouldn't take a dependency on it here. I don't think it's worth it in this context, particularly because creating the converted name only happens once and then the result is stored.

Though I believe it is feasible to do, if we really wanted to. For example:
https://github.com/dotnet/corefx/blob/c78b7189eabe20c6f4c8d386ca18b68fea513189/src/System.Runtime.Extensions/src/System/Net/WebUtility.cs#L58

The main benefit would be on startup, but we might be able to write this method without using `ValueStringBuilder` and still be faster. I am heads down on some other work currently, and will review this change soon. Thanks for your patience, @YohDeadfall, appreciate it.

Reply:
> I don't think it's worth it in this context, particularly because creating the converted name only happens once and then the result is stored

@ahsonkhan, are there circumstances where it might happen for every occurrence? e.g. when happens when serializing a dictionary?

Reply:
I will take a second look and verify, but from what I understand, calls to `ConvertName` should only ever happen once. @steveharter, @layomia - feel free to chime in about the behavior of serializing dictionary keys/values.

Reply:
> from what I understand, calls to ConvertName should only ever happen once

That doesn't seem to be the case, e.g.
https://github.com/dotnet/corefx/blob/1338e4e8a542c3b3e20d99e544e09860f23e99ea/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs#L154
but maybe I'm just missing where that gets cached.  I wouldn't expect it to be cached, though, considering arbitrary dictionaries might get serialized, and their keys could be all over the map.

Reply:
Good point @stephentoub. I was looking at `PropertyNamingPolicy` only, but `JsonNamingPolicy` applies to `DictionaryKeyPolicy` as well.

Once the implementation of `JsonSnakeCaseNamingPolicy` is well tested, we should consider trying to optimize the built-in naming policies. This can be done in a follow-up PR (no point trying to optimize it without a larger test set that ensures correctness). Filed an issue: https://github.com/dotnet/corefx/issues/42581

``` ini

BenchmarkDotNet=v0.11.5.1159-nightly, OS=Windows 10.0.18362
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.100-preview1-014459
  [Host]     : .NET Core 3.0.0 (CoreCLR 4.700.19.46205, CoreFX 4.700.19.46214), X64 RyuJIT
  Job-HPZZRZ : .NET Core 3.0.0 (CoreCLR 4.700.19.46205, CoreFX 4.700.19.46214), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  IterationTime=250.0000 ms  MaxIterationCount=10  
MinIterationCount=5  WarmupCount=1  

```
|  Method |     Mean |     Error |    StdDev |   Median |      Min |      Max | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|--------:|-------:|------:|------:|----------:|
| Default | 2.426 us | 0.0473 us | 0.0281 us | 2.425 us | 2.391 us | 2.473 us |  1.00 |    0.00 | 0.1567 |     - |     - |     656 B |
|   Camel | 3.040 us | 0.0895 us | 0.0592 us | 3.027 us | 2.979 us | 3.150 us |  1.26 |    0.04 | 0.2491 |     - |     - |    1048 B |
|   Snake | 4.293 us | 0.0939 us | 0.0621 us | 4.280 us | 4.195 us | 4.397 us |  1.77 |    0.01 | 0.5624 |     - |     - |    2432 B |


```C#
public class SerializeDictionaryWithNamingPolicy
{
    JsonSerializerOptions _camelOptions;
    JsonSerializerOptions _snakeOptions;

    Dictionary<string, string> _foo;

    [GlobalSetup]
    public void Setup()
    {
        _camelOptions = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase, DictionaryKeyPolicy = JsonNamingPolicy.CamelCase };
        _snakeOptions = new JsonSerializerOptions { PropertyNamingPolicy = new JsonSnakeCaseNamingPolicy(), DictionaryKeyPolicy = new JsonSnakeCaseNamingPolicy() };

        _foo = new Dictionary<string, string>()
        {
            ["ID"] = "A",
            ["url"] = "A",
            ["URL"] = "A",
            ["THIS IS SPARTA"] = "A",
            ["IsCIA"] = "A",
            ["iPhone"] = "A",
            ["IPhone"] = "A",
            ["xml2json"] = "A",
            ["already_snake_case"] = "A",
            ["IsJSONProperty"] = "A",
            ["ABCDEFGHIJKLMNOP"] = "A",
            ["Hi!! This is text.Time to test."] = "A",
        };
    }

    [BenchmarkCategory(Categories.CoreFX, Categories.JSON)]
    [Benchmark(Baseline = true)]
    public string Default()
    {
        return JsonSerializer.Serialize(_foo);
    }

    [BenchmarkCategory(Categories.CoreFX, Categories.JSON)]
    [Benchmark]
    public string Camel()
    {
        return JsonSerializer.Serialize(_foo, _camelOptions);
    }

    [BenchmarkCategory(Categories.CoreFX, Categories.JSON)]
    [Benchmark]
    public string Snake()
    {
        return JsonSerializer.Serialize(_foo, _snakeOptions);
    }
}
```